### PR TITLE
Globalize amd support

### PR DESCRIPTION
**Pull request by treshugart**

### CODE REVIEW

File: src/transform/globalize.js [hunk 1]
```diff
@@ -29,30 +28,93 @@ function indent (code) {
...
+  }));
+
+  if (rval) {
+    exports = module.exports = rval;
+  }
+}
+
 module.exports = function () {
   return function (data, info) {
-    var sourcemap = data.match(regexSourcemap);
```
Comment (treshugart) [inline src/transform/globalize.js]: This was leaked from a previous commit. Didn't affect the code, but removing because it will be implemented in a separate PR.

File: src/transform/globalize.js [hunk 2]
```diff
@@ -29,30 +28,93 @@ function indent (code) {
...
+  });
+
+  keyVals.unshift('"exports": exports');
+  keyVals.unshift('"module": module');
+  code = '\n' + indent(keyVals.join(',\n')) + '\n';
+  return '{' + code + '}';
+}
+
+function defineReplacement (name, deps, func) {
+  function overload (type) {
```
Comment (treshugart) [inline src/transform/globalize.js]: The code to overload amd arguments would be about this much anyways. Was hoping to abstract it out at some point. Maybe this could be made into a more concise block instead. Thoughts?

File: src/transform/globalize.js [hunk 3]
```diff
@@ -29,30 +28,75 @@ function indent (code) {
...
-    data = 'var module = { exports: {} };\nvar exports = module.exports;\n\n' + data;
+    // We assume CommonJS because that's what we're using to convert it.
+    vars.push('var module = {\n' + indent('exports: {}') + '\n};');
+    vars.push('var exports = module.exports;');
+
+    // We only need to generate the AMD -> CommonJS shim if it's used.
+    if (isAmd) {
+      vars.push('var defineDependencies = ' + defineDependencies(info.imports, info.dependencies) + ';');
+      vars.push('var define = ' + defineReplacement + ';');
+      vars.push('define.amd = true;');
```
Comment (cyberdash) [inline src/transform/globalize.js]: Maybe the name of this can be changed since it's not just `vars` now? Something like `injectedLines` / `shim`?

File: src/transform/globalize.js [hunk 4]
```diff
@@ -29,30 +28,75 @@ function indent (code) {
...
+}
+
+function defineReplacement (name, deps, func) {
+  var rval;
+  func = [func, deps, name].filter(function (cur) { return typeof cur === 'function'; })[0];
+  deps = [deps, name, []].filter(Array.isArray)[0];
+  rval = func.apply(null, deps.map(function (value) {
+    return defineDependencies[value];
+  }));
+  return rval && (exports = module.exports = rval);
```
Comment (cyberdash) [inline src/transform/globalize.js]: Maybe this is an edge case, but if something falsey was returned from `rval` (zero?), `module.exports` would be undefined rather than zero (I think?)

**Comment by cyberdash** [comment]
that's all of my comments! Looks good to me :)

File: src/transform/globalize.js (hunk 4 above)
Comment (treshugart) [inline src/transform/globalize.js]: When using Babel, it doesn't check it's type and immediately tries to call `Object.defineProperty` on it. Perhaps we should be a bit more stringent here.

File: src/transform/globalize.js (hunk 3 above)
Comment (treshugart) [inline src/transform/globalize.js]: Good idea.
